packages/hierarchies-react/src/presentation-hierarchies-react/itwinui/TreeRenderer.tsx

```diff
@@ -45,16 +46,50 @@ export function TreeRenderer({ rootNodes, expandNode, localizedStrings, selectNo
     selectNodes: selectNodes ?? noopSelectNodes,
     selectionMode: selectionMode ?? "single",
   });
-
+  const parentRef = useRef<HTMLDivElement>(null);
   const flatNodes = useMemo(() => flattenNodes(rootNodes), [rootNodes]);
 
+  const virtualizer = useVirtualizer({
+    count: flatNodes.length,
+    getScrollElement: () => parentRef.current,
+    estimateSize: () => 28,
+    overscan: 5,
+  });
+
+  const items = virtualizer.getVirtualItems();
+
   return (
     <LocalizationContextProvider localizedStrings={localizedStrings}>
-      <Tree.Root style={{ height: "100%", width: "100%" }}>
-        {flatNodes.map((flatNode) => (
-          <TreeNodeRenderer {...treeProps} expandNode={expandNode} onNodeClick={onNodeClick} onNodeKeyDown={onNodeKeyDown} node={flatNode} key={flatNode.id} />
-        ))}
-      </Tree.Root>
+      <div style={{ height: "100%", width: "100%", overflowY: "auto" }} ref={parentRef}>
+        <Tree.Root style={{ height: virtualizer.getTotalSize(), width: "100%", position: "relative" }}>
+          {items.map((virtualizedItem) => {
+            return (
+              <div
```

Comment:
`@itwin/itwinui-react@5.0.0-alpha.7` passthrough `style` props to the underlying DOM element used to render `Tree.Item` so you can remove this intermediate `div`.

Reply:
saw that already, thinking of removing measure logic for now from virtualization to avoid readding ref passing into node renderer as right now we do not have sublabels, and all nodes are the same size.

Reply:
I think we should readd ref forwarding in this PR so virtualization would be implemented fully and we won't need to get back to it later.

Reply:
Readded refs, as well as removed wrapper around tree.item